src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala

```diff
@@ -34,100 +38,165 @@ object StructuralVariant {
-      if (!r.isMateMapped) {
-        Seq[Long]()
-      } else {
-        val mateStart = r.mateAlignmentProperties.map(_.start).getOrElse(r.read.start)
-        val start = Math.min(r.read.start, mateStart)
-        val roundedStart = start / BLOCK_SIZE * BLOCK_SIZE
-        val insertSize = orientedInsertSize(r).getOrElse(0)
-        roundedStart to (roundedStart + insertSize) by BLOCK_SIZE
+    case class ExceptionalReadsReturnType(
+      val readsInRange: RDD[PairedMappedRead],
+      val insertSizes: RDD[Int],
+      val insertStats: MedianStats,
+      val maxNormalInsertSize: Int,
+      val exceptionalReads: RDD[PairedMappedRead])
+
+    def getExceptionalReads(reads: RDD[PairedMappedRead]): ExceptionalReadsReturnType = {
+      // Pare down to mate pairs which aren't highly displaced & aren't inverted or translocated.
+      val readsInRange = reads.filter(r => {
+        val s = r.inferredInsertSize
+        r.read.referenceContig == r.mate.referenceContig &&
+          r.read.isPositiveStrand != r.mate.isPositiveStrand &&
+          s < MAX_INSERT_SIZE
+      })
+      readsInRange.persist()
+
+      val insertSizes = readsInRange.map(orientedInsertSize)
+      val insertStats = medianStats(insertSizes.takeSample(false, 100000))
+      println("insert stats: " + insertStats)
+
+      // Find blocks where the average insert size differs significantly from the genome-wide median.
+      // Following DELLY, we use a threshold of 5 median absolute deviations above the median.
+      val maxNormalInsertSize = (insertStats.median + 5 * insertStats.mad).toInt
+
+      val exceptionalReads = readsInRange.filter(_.inferredInsertSize > maxNormalInsertSize)
+
+      ExceptionalReadsReturnType(
+        readsInRange,
+        insertSizes,
+        insertStats,
+        maxNormalInsertSize,
+        exceptionalReads
+      )
+    }
+
+    // Given two pairs of reads, is there a deletion which would make both of them have normal insert sizes?
+    // TODO: this assumes that both reads have read.start.pos < read.mate.pos
+    def areReadsIncompatible(read1: PairedMappedRead, read2: PairedMappedRead, maxNormalInsertSize: Long): Boolean = {
+      if (read1.read.start < read2.read.start) return areReadsIncompatible(read2, read1, maxNormalInsertSize)
+
+      // This matches the DELLY logic; the last five lines are copy/paste. See
+      // https://github.com/tobiasrausch/delly/blob/7464cacfe1d420ac5bd7ed40a3093a80a93a6964/src/tags.h#L441-L445
+      val pair1Min = Math.min(read1.read.start, read1.mate.start)
+      val pair1Max = pair1Min + read1.inferredInsertSize
+      val pair1ReadLength = read1.read.sequence.length
+
+      val pair2Min = Math.min(read2.read.start, read2.mate.start)
+      val pair2Max = pair2Min + read2.inferredInsertSize
+      val pair2ReadLength = read2.read.sequence.length
+
+      if ((pair2Min + pair2ReadLength - pair1Min) > maxNormalInsertSize) return true;
```

Comment:
can you comment these lines explaining what they do? the lack of obvious a pair1⟺pair2 symmetry is disconcerting.